pageserver/src/tenant/secondary/heatmap_uploader.rs

```diff
@@ -0,0 +1,582 @@
+                last_upload: None,
+                next_upload: Some(Instant::now()),
+                last_digest: None,
+            });
+
+        // Decline to do the upload if insufficient time has passed
+        if state.next_upload.map(|nu| &nu > now).unwrap_or(false) {
+            return;
+        }
+
+        let last_digest = state.last_digest;
+        self.tenants_pending.push_back(UploadPending {
+            tenant,
+            last_digest,
+        })
+    }
+
+    fn spawn_pending(&mut self) {
+        while !self.tenants_pending.is_empty()
+            && self.tenants_uploading.len() < self.concurrent_uploads
+        {
+            // unwrap: loop condition includes !is_empty()
+            let pending = self.tenants_pending.pop_front().unwrap();
+            self.spawn_upload(pending.tenant, pending.last_digest);
+        }
+    }
+
+    fn spawn_upload(&mut self, tenant: Arc<Tenant>, last_digest: Option<md5::Digest>) {
+        let remote_storage = self.remote_storage.clone();
+        let tenant_shard_id = *tenant.get_tenant_shard_id();
+        let (completion, barrier) = utils::completion::channel();
+        let result_tx = self.task_result_tx.clone();
+        self.tasks.spawn(async move {
+            // Guard for the barrier in [`WriteInProgress`]
+            let _completion = completion;
+
+            let started_at = Instant::now();
+            let digest = match upload_tenant_heatmap(remote_storage, &tenant, last_digest).await {
+                Ok(UploadHeatmapOutcome::Uploaded(digest)) => {
+                    let duration = Instant::now().duration_since(started_at);
+                    SECONDARY_MODE
+                        .upload_heatmap_duration
+                        .observe(duration.as_secs_f64());
+                    SECONDARY_MODE.upload_heatmap.inc();
+                    Some(digest)
+                }
+                Ok(UploadHeatmapOutcome::NoChange | UploadHeatmapOutcome::Skipped) => last_digest,
+                Err(UploadHeatmapError::Upload(e)) => {
+                    tracing::warn!(
+                        "Failed to upload heatmap for tenant {}: {e:#}",
+                        tenant.get_tenant_shard_id(),
+                    );
+                    let duration = Instant::now().duration_since(started_at);
+                    SECONDARY_MODE
+                        .upload_heatmap_duration
+                        .observe(duration.as_secs_f64());
+                    SECONDARY_MODE.upload_heatmap_errors.inc();
+                    last_digest
+                }
+                Err(UploadHeatmapError::Cancelled) => {
+                    tracing::info!("Cancelled heatmap upload, shutting down");
+                    last_digest
+                }
+            };
+
+            let now = Instant::now();
+            let next_upload = tenant
+                .get_heatmap_period()
+                .and_then(|period| now.checked_add(period));
```

Comment:
This was intentional -- because heatmap uploads are a low priority task, I prefer to have the implicit behavior of waiting longer between uploads if they're taking a long time.  Until we have a more concrete system of prioritizing jobs/IOs, little hacks like this help to bias the behavior of the system toward doing less work for heatmap uploads and more of the higher priority work.

Bear in mind that the expected interval for heatmap uploads is multiple minutes, so all the timing is very lazy here.

>So, I think what we want is the equivalent of tokio::time::interval + [MissedTickBehavior::Skip](https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html#variant.Skip) + make noise if we miss (like warn_when_period_overrun does).

Let's pick this up in the same context as the point about warning on late execution (https://github.com/neondatabase/neon/pull/6050#discussion_r1426571754).  This scheduler isn't really intented to provide guarantees of on-time execution, but it will be worth defining some bound (e.g. aim to run within a minute of the schedule).  That same bound could limit how much we will adjust next execution time to account how far behind schedule we are, so that something scheduled for every 10 minutes really will run every 10 minutes, but if stuff is running well behind schedule we won't try and run the next iteration excessively soon.